bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/api/utils/TypeUtils.java

```diff
@@ -55,23 +55,27 @@ private TypeUtils() {
     }
 
     public static boolean isValueType(Type type) {
-        if (type == TYPE_INT || type == TYPE_BYTE ||
-                type == TYPE_FLOAT ||
-                type == TYPE_DECIMAL || type == TYPE_STRING ||
-                type == TYPE_BOOLEAN) {
-            return true;
-        }
-
-        if (type != null && type.getTag() == TypeTags.FINITE_TYPE_TAG) {
-            // All the types in value space should be value types.
-            for (Object value : ((BFiniteType) type).valueSpace) {
-                if (!isValueType(TypeChecker.getType(value))) {
-                    return false;
+        switch (type.getTag()) {
+            case TypeTags.INT_TAG:
+            case TypeTags.BYTE_TAG:
+            case TypeTags.FLOAT_TAG:
+            case TypeTags.DECIMAL_TAG:
+            case TypeTags.BOOLEAN_TAG:
+            case TypeTags.STRING_TAG:
+                return true;
+            case TypeTags.FINITE_TYPE_TAG:
+                for (Object value : ((BFiniteType) type).valueSpace) {
+                    if (!isValueType(TypeChecker.getType(value))) {
+                        return false;
+                    }
                 }
-            }
-            return true;
+                return true;
+            case TypeTags.TYPE_REFERENCED_TYPE_TAG:
+                return isValueType(((ReferenceType) type).getReferredType());
```

Comment:
Why do we handle this in a separate case instead of doing `getReferredType()` in the first place?

Reply:
The implementation of `TypeUtils.getReferredType()` contains an `if` condition recursively. By going through the `switch-case`, we can improve the execution if the `type` argument is not a `ReferenceType`.

Reply:
If it is not a `ReferenceType` then `TypeUtils.getReferredType()` wont go recursively right? It will just return the type